src/APIBasedAuth.ts

```diff
@@ -0,0 +1,159 @@
+export const formatAxiosError = <ErrorResponse = any>(
+  error: AxiosError<ErrorResponse> | unknown
+) => {
+  if (!axios.isAxiosError(error)) {
+    throw new AxiosError(String(error));
+  }
+  throw error.response?.data;
+};
+
+class APIBasedAuth {
+  private client: AxiosInstance;
+  private clientOptions: APIBasedAuth.Config;
+  private session?: APIBasedAuth.Session;
+
+  constructor({ clientId, ...options }: APIBasedAuth.Config) {
+    if (!clientId) {
+      throw new Error('APIBasedAuth param clientId is required');
+    }
+
+    this.client = axios.create({
+      ...options,
+      baseURL: options.baseURL || DEFAULT_BASE_URL
+    });
+
+    this.clientOptions = { clientId, ...options };
+  }
+
+  public confirmPasswordlessAuth(
+    baseInput: Omit<APIBasedAuth.VerifyPasswordlessAuthData, 'clientId'>
+  ) {
+    const input = {
+      ...baseInput,
+      // if session or username is not present get stored properties
+      ...(!baseInput.session || !baseInput.username ? this.session : null)
+    };
+    return this.client
+      .post<
+        APIBasedAuth.SuccessfulAuthResponse,
+        AxiosResponse<APIBasedAuth.SuccessfulAuthResponse>,
+        APIBasedAuth.VerifyPasswordlessAuthData
+      >('/passwordless-auth/verify', {
+        clientId: this.clientOptions.clientId,
+        code: input.code,
+        session: input.session,
+        username: input.username
+      })
+      .then(({ data }) => {
+        // TODO: add storage
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  public initiatePasswordAuth(
+    input: Omit<APIBasedAuth.SignInData, 'clientId'>
+  ) {
+    return this.client
+      .post<
+        APIBasedAuth.SuccessfulAuthResponse,
+        AxiosResponse<APIBasedAuth.SuccessfulAuthResponse>,
+        APIBasedAuth.SignInData
+      >('/login', {
+        clientId: this.clientOptions.clientId,
+        ...input
+      })
+      .then(({ data }) => {
+        // TODO: add storage
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  public initiatePasswordlessAuth({
+    appsBaseUri,
+    loginAppBasePath,
+    username
+  }: Omit<APIBasedAuth.PasswordlessAuthData, 'clientId'>) {
+    return this.client
+      .post<
+        APIBasedAuth.PasswordlessAuthResponse,
+        AxiosResponse<APIBasedAuth.PasswordlessAuthResponse>,
+        APIBasedAuth.PasswordlessAuthData
+      >('/passwordless-auth', {
+        appsBaseUri,
+        clientId: this.clientOptions.clientId,
+        loginAppBasePath,
+        username
+      })
+      .then(({ data }) => {
+        this.session = {
+          session: data.session,
+          username
+        };
+        return data;
+      })
+      .catch(formatAxiosError);
+  }
+
+  public logout() {
+    this.session = undefined;
+  }
+}
+
+declare namespace APIBasedAuth {
+  export type Config = Pick<
+    AxiosRequestConfig,
+    'baseURL' | 'withCredentials'
+  > & {
+    clientId: string;
+  };
```

Comment:
Two questions:
1. Why is `withCredentials` included here?
2. Can we either:
  - avoid picking from the `AxiosRequestConfig` type, or
  - include `@types/axios` in our `dependencies` array?

If we export references to AxiosRequestConfig from our type declarations, type compilation will fail when someone tries to use this without `@types/axios` in their dependencies. It should go in ours.

Reply:
1. it's there cause I see we're using it for preventia in ClientUI
2. yeah, I'll just type it individually instead of adding the lib

Reply:
> it's there cause I see we're using it for preventia in ClientUI

You're right that the flag is specified, but it's statically set to `false` for the client that interacts with the auth api. `false` is the default setting -- we should be able to leave it off here.

@jkdowdle can confirm

Reply:
I'll change that in my follow up with storage